### ISSUE TITLE
[HOLD Bedrock-PHP 218 and Web-E 44934] Search - Deleted search doesn´t disappear from secondary device despite refreshing page

### ISSUE THREAD

**Issue by vincdargento** [Expensify/App#53939]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.74-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/5328850&group_by=cases:section_id&group_order=asc&group_id=296775
**Email or phone of affected tester (no customers):** ibellicotest+88@gmail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
Prerequisite: Account has to be opened in two devices or environments at the same time.
1. Open the Expensify app.
2. Tap on "Search" on the bottom of the screen.
3. Tap on the filters icon.
4. Apply any filter.
5. Tap on "Save Search"
6. Switch to second device and verify that the search was created too.
7. On main device, tap on dropdown menu.
8. Tap on the three dots icon on the saved search and delete it.
9. Switch to second device again and refresh the page.

## Expected Result:
Search deleted from main device, should disappear from every device in which the account is logged.

## Actual Result:
Deleted saved search, doesn´t disappear from secondary device despite refreshing the page, and the user can interact with it normally.

## Workaround:
Unknown

## Platforms:
- [x] Android: Standalone
- [x] Android: HybridApp
- [x] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/9ec420f5-1500-4402-9fb6-0655b68731c2

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021866860154375173814</li>
        <li>Upwork Job ID: 1866860154375173814</li>
        <li>Last Price Increase: 2024-12-18</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @luacmartins</details>

**Comment by FitseTLT** [comment]
Already being handled in https://github.com/Expensify/App/issues/53687 Let's close it

**Comment by rojiphil** [comment]
> Already being handled in #53687 Let's close it

@FitseTLT Can you please explain how this issue is been handled in the mentioned issue?

**Comment by rojiphil** [comment]
@bfitzexpensify I think this issue is different. Can you please reopen this issue until [this comment](https://github.com/Expensify/App/issues/53939#issuecomment-2536516977) is resolved?

**Comment by FitseTLT** [comment]
The RCA is the same @rojiphil we re-request search action when the transaction length increase but in both cases the expense is deleted which means transaction length decreases hence we don't request search here
https://github.com/Expensify/App/blob/b2dca65a743aea0c09cbb5b711e832f2fc9a6e1a/src/hooks/useSearchHighlightAndScroll.ts#L46-L55
@bfitzexpensify please close it they are the same issues

**Comment by rojiphil** [comment]
> The RCA is the same @rojiphil we re-request search action when the transaction length increase but in both cases the expense is deleted which means transaction length decreases hence we don't request search here

@FitseTLT Wondering how the RCA can be the same when this issue is not about deleting an expense. Rather, it is about deleting the saved search. Are they not different?

**Comment by FitseTLT** [comment]
Wow very sorry @rojiphil 🙏 It's totally different issue. Sorry for the confusion here.

**Comment by MrMuzyk** [comment]
I am Michał from Callstack - expert contributor group. I’d like to work on this job.

**Comment by MrMuzyk** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When deleting last saved search it's not being removed correctly on parallel device

### What is the root cause of that problem?

This issue only happens for the last saved search that user tries to remove. If there are any saved searches left after delete everything works fine.

I believe the issues lays on the BE side. When request is processed on the "main device", BE sends a pusher update to sync all the connected clients for the user that made the request. However that pusher event is sent with an empty array `[]` instead of an empty object `{}` which causes fail during update inside Onyx on the "secondary device".

<img width="850" alt="Screenshot 2024-12-16 at 15 53 29" src="https://github.com/user-attachments/assets/aa75a4bf-7e3e-4aac-830b-0be06b5890e2" />

To confirm that, I've logged the value that this particular Onyx.set is receiving - was `[]`. I've also added a condition to modify passed value in that one scenario. When swapped to empty object `{}` everything executed without any issues and saved search was deleted on both accounts simultaneously

Modification to Onyx.set
```
function set(key, value) {
    console.log(key, value, 'onyxSet')
    if (key === 'nvp_savedSearches') {
        value = {};
    }

```

<img width="851" alt="Screenshot 2024-12-16 at 15 59 18" src="https://github.com/user-attachments/assets/4553c46a-d77e-4aef-8566-bb131deb9aae" />


### What changes do you think we should make in order to solve the problem?


I believe the issue lies on the BE side. When `DeleteSavedSearch` command that removes last saved search is processed by BE, pusher update that is being send should clear the `nvp_savedSearches` via empty object `{}` instead of array `[]`

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?

N/A

### What alternative solutions did you explore? (Optional)
 
 N/A

**Comment by rojiphil** [comment]
> I believe the issue lies on the BE side.

@MrMuzyk Thanks for the proposal. I agree with the [analysis](https://github.com/Expensify/App/issues/53939#issuecomment-2545941072). 
Let's bring in the internal engineer to weigh in here.
🎀👀🎀 C+ reviewed

**Comment by rojiphil** [comment]
Unassigning myself as there is nothing to be done from FE side. Will stay subscribed though. Thanks.

**Comment by carlosmiceli** [comment]
I have quite a bit on my plate so will have to unassign myself, maybe @luacmartins can fix this one quickly since he has a lot of context on Search.

**Comment by luacmartins** [comment]
I have a draft PR up, but I'm having issues since the object -> array conversion is still happening somewhere else. I'll keep investigating.

**Comment by luacmartins** [comment]
We couldn't figure out where the conversion is happening. We're instead gonna change the way our API parses json.

**Comment by luacmartins** [comment]
Putting this issue on hold for https://github.com/Expensify/Bedrock-PHP/pull/218 and https://github.com/Expensify/Web-Expensify/pull/44934

**Comment by luacmartins** [comment]
This is fixed now. Closing.
